### PR TITLE
Fix primary key checks

### PR DESCRIPTION
**Pull request by superkhau**
Connect to #706

### CODE REVIEW

File: lib/model-definition.js [hunk 1]
```diff
@@ -303,24 +303,6 @@ ModelDefinition.prototype.toJSON = function (forceRebuild) {
...
-
-  // check if id is defined by a relation
-  this.ids().forEach(function(item) {
-    if (item.hasOwnProperty('id'))
-      isDefined = true;
-  });
-
-  return isDefined;
+ModelDefinition.prototype.hasPkDefined = function() {
+  return (this.ids().length !== 0 || this.idNames().length !== 0);
```
Comment (raymondfeng) [inline lib/model-definition.js]: No need to check both. Either ids() or idNames() is good enough.

File: lib/model-definition.js [hunk 2]
```diff
@@ -303,24 +303,6 @@ ModelDefinition.prototype.toJSON = function (forceRebuild) {
...
-
-  // check if id is defined by a relation
-  this.ids().forEach(function(item) {
-    if (item.hasOwnProperty('id'))
-      isDefined = true;
-  });
-
-  return isDefined;
+ModelDefinition.prototype.hasPrimaryKey = function() {
+  return (this.ids().length > 0 || this.idNames().length > 0);
```
Comment (superkhau) [inline lib/model-definition.js]: @raymondfeng Is this ok?

Comment (raymondfeng) [inline lib/model-definition.js]: It's good enough to have `return this.ids().length > 0;`

**Comment by superkhau** [comment]
@raymondfeng Review please.

File: lib/dao.js [hunk 3]
```diff
@@ -2542,13 +2550,8 @@ jutil.mixin(DataAccessObject, Relation);
...
-/**
- * Return an error object stating the `id` property must be part of the model
- * definition.
- *
- * @return {Object} The error stating the requirements
- */
-function getIdDefinitionError() {
-  return new Error('`id` property must be defined as part of the model ' +
-    'definition');
+function PrimaryKeyMissingError(msg) {
```
Comment (raymondfeng) [inline lib/dao.js]: We probably need to know which model it is.

File: lib/dao.js [hunk 4]
```diff
@@ -2542,13 +2550,9 @@ jutil.mixin(DataAccessObject, Relation);
...
- * definition.
- *
- * @return {Object} The error stating the requirements
- */
-function getIdDefinitionError() {
-  return new Error('`id` property must be defined as part of the model ' +
-    'definition');
+function PrimaryKeyMissingError(modelName) {
+  this.name = 'PrimaryKeyMissingError';
+  this.message = 'Primary key missing from the ' + modelName + ' model ' +
```
Comment (superkhau) [inline lib/dao.js]: @raymondfeng Including model name in the error message now.

File: lib/dao.js [hunk 5]
```diff
@@ -1839,7 +1839,9 @@ DataAccessObject.prototype.save = function (options, cb) {
   assert(typeof options === 'object', 'The options argument should be an object');
   assert(typeof cb === 'function', 'The cb argument should be a function');
 
-  if (this.isNewRecord()) {
+  if (!Model.definition.hasPrimaryKey()) {
+    process.nextTick(cb(new PrimaryKeyMissingError(Model.modelName)));
```
Comment (raymondfeng) [inline lib/dao.js]: Do you need to `return cb.promise;`?

File: lib/dao.js [hunk 6]
```diff
@@ -2141,6 +2143,9 @@ DataAccessObject.prototype.remove =
       var id = getIdValue(this.constructor, this);
       var hookState = {};
 
+      if (!Model.definition.hasPrimaryKey())
+        process.nextTick(cb(new PrimaryKeyMissingError(Model.modelName)));
+
```
Comment (raymondfeng) [inline lib/dao.js]: Do you need to `return cb.promise;`?

File: lib/dao.js [hunk 7]
```diff
@@ -2327,6 +2332,9 @@ DataAccessObject.prototype.updateAttributes = function updateAttributes(data, op
   assert(typeof connector.updateAttributes === 'function',
     'updateAttributes() must be implemented by the connector');
 
+  if (!Model.definition.hasPrimaryKey())
+    process.nextTick(cb(new PrimaryKeyMissingError(Model.modelName)));
+
```
Comment (raymondfeng) [inline lib/dao.js]: Do you need to `return cb.promise;`?

File: lib/dao.js (hunk 5 above)
Comment (superkhau) [inline lib/dao.js]: Actually, yes I think so.

File: lib/dao.js [hunk 8]
```diff
@@ -856,8 +856,9 @@ DataAccessObject.findById = function findById(id, filter, options, cb) {
   assert(typeof options === 'object', 'The options argument must be an object');
   assert(typeof cb === 'function', 'The cb argument must be a function');
 
-  if (!this.definition.hasIdDefined()) {
-    process.nextTick(cb(getIdDefinitionError()));
+  if (!this.definition.hasPrimaryKey()) {
+    process.nextTick(cb(new PrimaryKeyMissingError(this.modelName)));
```
Comment (raymondfeng) [inline lib/dao.js]: This is not going to work. `process.nextTick()` takes a function. The code should look like the following:
```js

   // Define a helper function
   function isPKMissing(modelClass, cb) {
     var hasPK = modelClass.definition.hasPrimaryKey();
     if (hasPK) return false;
     process.nextTick(function() {
       cb(new PrimaryKeyMissingError(modelClass.modelName);
     }
     return true;
 }

// Use the helper function
if (isPKMissing(this, cb)) return cb.promise;
```

File: lib/dao.js [hunk 9]
```diff
@@ -856,8 +856,8 @@ DataAccessObject.findById = function findById(id, filter, options, cb) {
   assert(typeof options === 'object', 'The options argument must be an object');
   assert(typeof cb === 'function', 'The cb argument must be a function');
 
-  if (!this.definition.hasIdDefined()) {
-    process.nextTick(cb(getIdDefinitionError()));
+  if (isPKNotFound(this, cb)) {
```
Comment (raymondfeng) [inline lib/dao.js]: Nick-pick: double-negative is confusing.

Comment (superkhau) [inline lib/dao.js]: True, I will change it all back to `Missing` ;)

**Comment by raymondfeng** [comment]
LGTM

**Comment by superkhau** [comment]
@slnode test please

**Pull request merged**